kroxylicious-kms-provider-aws-kms-test-support/src/main/java/io/kroxylicious/kms/provider/aws/kms/AbstractAwsKmsTestKmsFacade.java

```diff
@@ -52,4 +96,202 @@ public final void stop() {
         stopKms();
     }
 
+    @Override
+    public TestKekManager getTestKekManager() {
+        return new AwsKmsTestKekManager();
+    }
+
+    class AwsKmsTestKekManager implements TestKekManager {
+        @Override
+        public void generateKek(String alias) {
+            Objects.requireNonNull(alias);
+
+            if (exists(alias)) {
+                throw new AlreadyExistsException(alias);
+            }
+            else {
+                create(alias);
+            }
+        }
+
+        @Override
+        public void rotateKek(String alias) {
+            Objects.requireNonNull(alias);
+
+            if (!exists(alias)) {
+                throw new UnknownAliasException(alias);
+            }
+            else {
+                rotate(alias);
+            }
+        }
+
+        @Override
+        public void deleteKek(String alias) {
+            if (!exists(alias)) {
+                throw new UnknownAliasException(alias);
+            }
+            else {
+                delete(alias);
+            }
+        }
+
+        @Override
+        public boolean exists(String alias) {
+            try {
+                read(alias);
+                return true;
+            }
+            catch (UnknownAliasException uae) {
+                return false;
+            }
+        }
+
+        private void create(String alias) {
+            final CreateKeyRequest createKey = new CreateKeyRequest("key for alias: " + alias);
+            var createRequest = createRequest(createKey, TRENT_SERVICE_CREATE_KEY);
+            var createKeyResponse = sendRequest(alias, createRequest, CREATE_KEY_RESPONSE_TYPE_REF);
+
+            final CreateAliasRequest createAlias = new CreateAliasRequest(createKeyResponse.keyMetadata().keyId(), AwsKms.ALIAS_PREFIX + alias);
+            var aliasRequest = createRequest(createAlias, TRENT_SERVICE_CREATE_ALIAS);
+            sendRequestExpectingNoResponse(aliasRequest);
+        }
+
+        private DescribeKeyResponse read(String alias) {
+            final DescribeKeyRequest describeKey = new DescribeKeyRequest(AwsKms.ALIAS_PREFIX + alias);
+            var request = createRequest(describeKey, TRENT_SERVICE_DESCRIBE_KEY);
+            return sendRequest(alias, request, DESCRIBE_KEY_RESPONSE_TYPE_REF);
+        }
+
+        private void rotate(String alias) {
+            var key = read(alias);
+            final RotateKeyRequest rotateKey = new RotateKeyRequest(key.keyMetadata().keyId());
+            var rotateKeyRequest = createRequest(rotateKey, TRENT_SERVICE_ROTATE_KEY);
+            try {
+                sendRequestExpectingNoResponse(rotateKeyRequest);
+            }
+            catch (IllegalStateException e) {
+                if (e.getMessage().contains("501")) { // only when StatusCode = 501
+                    rotateInLocalStack(alias);
```

Comment:
That's the reason why I tried to override the `rotate` method in the `KubeAwsCloudKmsFacade` because it is the only one who implements this differently. But I agreed with Keith that it wasn't the right place to do it, as the right way to rotate is using the `RotateKeyOnDemand` and as a workaround we could use the pseudo rotation implementation done for `LocalStack`.

So let's name it `pseudoRotate` and call it when the usual rotation fails.